jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,175 @@
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/** Checks if Jib is up-to-date. */
+public class UpdateChecker {
+
+  /** JSON template for the configuration file used to enable/disable update checks. */
+  @VisibleForTesting
+  static class ConfigJsonTemplate implements JsonTemplate {
+    private boolean disableUpdateCheck;
+
+    @VisibleForTesting
+    void setDisableUpdateCheck(boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param skip if {@code true}, the update check itself will be skipped
+   * @param version the current version of Jib being used
+   * @param versionUrl the location to check for the latest version
+   * @param executorService the {@link ExecutorService}
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(
+      boolean skip, String version, String versionUrl, ExecutorService executorService) {
+    return new UpdateChecker(
+        executorService.submit(
+            () -> performUpdateCheck(skip, version, versionUrl, getConfigDir())));
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      boolean skip, String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (skip || Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    try {
+      Path configFile = configDir.resolve("config.json");
+      if (Files.exists(configFile)) {
+        // Abort of update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        JsonTemplateMapper.writeTo(config, Files.newOutputStream(configFile));
+      }
+
+      // Check time of last update check
+      Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+      if (Files.exists(lastUpdateCheck)) {
+        String fileContents =
+            new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+        try {
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
+        }
+      }
+
+      // Check for update
+      URLConnection connection = new URL(versionUrl).openConnection();
```

Comment:
This is not really necessary given the circumstances, but since it's not a big deal to do it, let's have `HttpURLConnection connection = (HttpURLConnection) ...` and call `connection.disconnect()` at the end.

And there should be some way to set some timeout values with `HttpURLConnection`. It's worth setting them a small value.

And can you verify Maven and Gradle both terminate right away even if this thread is alive, e.g., by putting `Thread.sleep(10000)` here?

Reply:
Looks like it does terminate right away.